src/Microsoft.VisualStudio.Composition/ComposedPart.cs

```diff
@@ -160,6 +160,57 @@ public IEnumerable<ComposedPartDiagnostic> Validate(IReadOnlyDictionary<Type, Ex
             }
         }
 
+        /// <summary>
+        /// Checks whether this part contains an import with <see cref="ImportCardinality.ExactlyOne"/> cardinality that has been invalidated,
+        /// which would invalidate this part as well.
+        /// </summary>
+        /// <param name="invalidPartDefnitionsSet">The set of definitions for the invalidated parts.</param>
+        /// <returns>The set of <see cref="ComposedPartDiagnostic"/> with the encountered errors.</returns>
+        public IEnumerable<ComposedPartDiagnostic> CheckForInvalidatedParts(ImmutableHashSet<ComposablePartDefinition> invalidPartDefnitionsSet)
+        {
+            foreach (var pair in this.SatisfyingExports)
+            {
+                var importDefinition = pair.Key.ImportDefinition;
+
+                bool invalidExportFound = false;
+
+                List<ExportDefinitionBinding> newSatisfyingExports = new List<ExportDefinitionBinding>();
```

Comment:
Instead of allocating a new collection when most of the time we won't use it, consider leaving it `null` and crafting the code below to initialize it and populate it only with the elements to be removed. That should improve perf and GC pressure.
You could then remove your `invalidExportFound` local since you can tell there's something to change by testing `newSatisfyingExports` (probably renamed) against `null`.